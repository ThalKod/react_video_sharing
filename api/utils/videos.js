const ffmpeg = require('fluent-ffmpeg');
const path = require("path");

const getDuration = (url, callback) =>{
  ffmpeg.ffprobe(url, (err, metadata) => {
    if(err) return callback({ error: true, msg: err});
    return callback({ error: false, duration: metadata.format.duration });
  });
};

const takeScreenshot = (video, duration, callback) => {
  const pathToScreenshot = path.join(__dirname, "../data", video.handle);
  let filename;
  const proc = ffmpeg(video.url)
      .on('filenames', (filenames) => {
        filename = filenames[0];
        console.log('filename : ' + filenames.join(', '));
      })
      .on('end', () => {
        console.log("Saved !");
        return callback({ error: false, pathToScreenshot: path.join(pathToScreenshot, filename) });
      })
      .on('error', (err) => {
        console.log("Error", err);
        return callback({ error: true, msg: err});
      })
      .takeScreenshots({ count: 1, timemarks: [ (duration / 4).toString() ], size: '1280x720' }, pathToScreenshot);
};

// process vidoe by taking a getting a video dutaion and a default screenscht
module.exports.processVideo = (video) => {
  return new Promise((resolve, reject) => {
    getDuration(video.url, (res) => {
      if(res.error) reject(res.msg);
      const duration = res.duration;
      takeScreenshot(video, duration, (result) => {
        if(result.error) reject(result.msg);
        const pathToScreenshot = result.pathToScreenshot;
        resolve({duration, pathToScreenshot});
      })
    })
  })
};
