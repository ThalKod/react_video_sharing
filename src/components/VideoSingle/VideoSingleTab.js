import React from "react";
import axios from "axios";
import { Link } from "react-router-dom"

import videoPlaceholder from "assets/images/video_placeholder.png";
import { request, formatSecondForRendering, formatCount } from "utils";


export default class VideoSingleTab extends React.Component{

  signal = axios.CancelToken.source();

  state = {
    videoImage: videoPlaceholder
  };

  componentDidMount = () => {
    const { id } = this.props;
    request("get", `/video/cover/default/${id}`, { cancelToken: this.signal.token })
        .then(res => {
          if(!res.data.error) this.setState({ videoImage: res.data.coverPhoto })
        })
        .catch(err => console.log(err));
  };

  componentWillUnmount = () => {
    this.signal.cancel('request canceled');
  };

  render(){
    const { name, duration, viewCount, id } = this.props;
    const { videoImage } = this.state;

    return(
        <div className="h-video row">
          <div className="col-lg-6 col-sm-6">
            <div className="v-img">
              <Link to={`/video/${id}`}><img src={videoImage} style={{ "border-radius": "5px"}} alt="cover" /></Link>
              <div className="time">{formatSecondForRendering(duration)}</div>
            </div>
          </div>
          <div className="col-lg-6 col-sm-6">
            <div className="v-desc">
              <Link to={`/video/${id}`}>{name}</Link>
            </div>
            <div className="v-views">{formatCount(viewCount)} Views</div>
          </div>
          <div className="clearfix"/>
        </div>
    );
  }
}



