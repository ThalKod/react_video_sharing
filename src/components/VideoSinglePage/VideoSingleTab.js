import React from "react";
import axios from "axios";
import { Link } from "react-router-dom"

// import VideoPlaceholder from "../../assets/images/video_placeholder.png";
import { request, formatSecondForRendering } from "../../utils";


export default class VideoSingleTab extends React.Component{

  signal = axios.CancelToken.source();

  state = {
    videoImage: ""
  };

  componentDidMount = () => {
    const { id } = this.props;
    request("get", `/video/cover/default/${id}`, { cancelToken: this.signal.token })
        .then(res => {
          if(!res.data.error) this.setState({ videoImage: res.data.coverPhoto })
        })
        .catch(err => console.log(err));
  };

  componentWillUnmount = () => {
    this.signal.cancel('request canceled');
  };

  render(){
    const { name, duration, viewCount, id } = this.props;
    const { videoImage } = this.state;

    return(
        <div className="h-video row">
          <div className="col-lg-6 col-sm-6">
            <div className="v-img">
              <Link to={`/video/${id}`}><img src={videoImage} alt="cover" /></Link>
              <div className="time">{formatSecondForRendering(duration)}</div>
            </div>
          </div>
          <div className="col-lg-6 col-sm-6">
            <div className="v-desc">
              <Link to={`/video/${id}`}>{name}</Link>
            </div>
            <div className="v-views">{viewCount} Views</div>
          </div>
          <div className="clearfix"/>
        </div>
    );
  }
}



