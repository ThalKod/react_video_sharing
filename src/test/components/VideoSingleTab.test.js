import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import videos from "test/__mock__/videos";
import VideoSingleTab from "components/VideoSingle/VideoSingleTab";

describe("VideoSingleTab", () => {
  it("should render the components correctly", () => {
    const props = { ...videos[0], id: "as0133"};
    const wrapper = shallow(<VideoSingleTab {...props}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  })
});
