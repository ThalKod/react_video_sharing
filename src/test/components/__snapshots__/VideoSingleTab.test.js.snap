// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoSingleTab should render the components correctly 1`] = `
<div
  className="h-video row"
>
  <div
    className="col-lg-6 col-sm-6"
  >
    <div
      className="v-img"
    >
      <Link
        replace={false}
        to="/video/as0133"
      >
        <img
          alt="cover"
          src=""
        />
      </Link>
      <div
        className="time"
      >
        3:40:42
      </div>
    </div>
  </div>
  <div
    className="col-lg-6 col-sm-6"
  >
    <div
      className="v-desc"
    >
      <Link
        replace={false}
        to="/video/as0133"
      >
        My Video
      </Link>
    </div>
    <div
      className="v-views"
    >
      25k
       Views
    </div>
  </div>
  <div
    className="clearfix"
  />
</div>
`;
